### PR TITLE
Add custom op for yolo post process

### PR DESCRIPTION
**Pull request by Zars19**
Using PopART API and codelets to form a custom operator running on IPU

### CODE REVIEW

File: ODLA/platforms/odla_popart/custom_ops/Makefile [hunk 1]
```diff
@@ -1,26 +1,26 @@
 CXX ?= g++
-CXXFLAGS = -std=c++14 -fPIC -g -DONNX_NAMESPACE=onnx -D_GLIBCXX_USE_CXX11_ABI=0
+CXXFLAGS = -std=c++14 -fPIC -g -DONNX_NAMESPACE=onnx -D_GLIBCXX_USE_CXX11_ABI=1
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/custom_ops/Makefile:2]: can we still use ABI=0 ?

File: ODLA/platforms/odla_popart/custom_ops/Makefile [hunk 2]
```diff
@@ -1,26 +1,26 @@
...
 
-rsqrt_custom_op: rsqrt.cc erf.cc
-	$(CXX) $(SOURCES)  $(LDLIBS) $(CXXFLAGS) $(INCLUDES) -o $(TARGET)
+custom_op:  erf.cc rsqrt.cc nms.cc 
+	$(CXX) $^ $(LDLIBS) $(CXXFLAGS) $(INCLUDES) -o $(TARGET)
 
-rsqrt_test: rsqrt_test.cc rsqrt_custom_op
-	$(CXX) -std=c++14 rsqrt_test.cc -lpopart -lpoplar -lpopops -ldl -DONNX_NAMESPACE=onnx -o rsqrt_test -D_GLIBCXX_USE_CXX11_ABI=0
+rsqrt_test: rsqrt_test.cc custom_op
+	$(CXX) -std=c++14 rsqrt_test.cc -lpopart -lpoplar -lpopops -ldl -DONNX_NAMESPACE=onnx -o rsqrt_test -D_GLIBCXX_USE_CXX11_ABI=1
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/custom_ops/Makefile:20]: we can use $(CXXFLAGS) here

File: ODLA/platforms/odla_popart/custom_ops/Makefile [hunk 3]
```diff
@@ -1,26 +1,26 @@
...
 
-erf_test: erf_test.cc rsqrt_custom_op
-	$(CXX) -std=c++14 erf_test.cc -lpopart -lpoplar -lpopops -ldl -DONNX_NAMESPACE=onnx -o erf_test -D_GLIBCXX_USE_CXX11_ABI=0
+erf_test: erf_test.cc custom_op
+	$(CXX) -std=c++14 erf_test.cc -lpopart -lpoplar -lpopops -ldl -DONNX_NAMESPACE=onnx -o erf_test -D_GLIBCXX_USE_CXX11_ABI=1
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/custom_ops/Makefile:23]: same here

File: models/vision/detection/yolo/Makefile [hunk 4]
```diff
@@ -7,25 +7,26 @@ SHELL := /bin/bash
 
 flags=-O3
 cv_libs:=-L/opt/opencv/3.4.11/lib64 -lopencv_core -lopencv_imgcodecs -lopencv_imgproc -lopencv_videoio
+CXXFLAGS = -std=c++14 -fPIC -g -DONNX_NAMESPACE=onnx -D_GLIBCXX_USE_CXX11_ABI=1
```
Comment (weimingzha0) [inline models/vision/detection/yolo/Makefile:10]: I think you don't need to set  CXX11_ABI here. ODLA APIs are C functions.

File: models/vision/detection/yolo/yolov3_postproc.cc [hunk 5]
```diff
@@ -15,14 +15,30 @@
 // limitations under the License.
 // =============================================================================
 
+#include <dlfcn.h>
```
Comment (weimingzha0) [inline models/vision/detection/yolo/yolov3_postproc.cc:18]: maybe we should make a separate version? This one is for CPU postprocessing.
You can make a yolov3_postproc_ipu.cc

Comment (weimingzha0) [reply models/vision/detection/yolo/yolov3_postproc.cc:18]: and here you shouldn't include IPU header files.
you can use ODLA APIs

Comment (weimingzha0) [reply models/vision/detection/yolo/yolov3_postproc.cc:18]: You can implement odla_NMS inside    [ODLA/platforms/odla_popart/odla_nn.cc](https://github.com/alibaba/heterogeneity-aware-lowering-and-optimization/blob/master/ODLA/platforms/odla_popart/odla_ops_nn.cc) 

Refer to [erf example](https://github.com/alibaba/heterogeneity-aware-lowering-and-optimization/blob/9a6a6831cc893b655e3ed30f240e813c07cba0f6/ODLA/platforms/odla_popart/odla_ops.cc#L102)

File: models/vision/detection/yolo/yolov3_postproc.cc [hunk 6]
```diff
@@ -96,56 +112,105 @@ std::vector<std::pair<std::string, std::array<float, 5>>> post_process_nhwc(
...
-        selected = i;
-        ret.push_back({ClassesNames[cls],
-                       {boxes[i][0], boxes[i][1], boxes[i][2], boxes[i][3],
-                        boxes[i][5 + cls]}});
-      }
+  void* handle = dlopen(
+
+      "/root/halo/ODLA/platforms/odla_popart/custom_ops/build/libcustom_ops.so",
+      RTLD_LAZY);
+  if (!handle) {
```
Comment (youbeny) [inline models/vision/detection/yolo/yolov3_postproc.cc:124]: we'd better use relative path

**Review by weimingzha0 — COMMENTED** [review]
btw, ODLA/platforms/odla_popart/custom_ops/build/libcustom_ops.so  should be removed

File: models/vision/detection/yolo/Makefile [hunk 7]
```diff
@@ -7,6 +7,7 @@ SHELL := /bin/bash
 
 flags=-O3
 cv_libs:=-L/opt/opencv/3.4.11/lib64 -lopencv_core -lopencv_imgcodecs -lopencv_imgproc -lopencv_videoio
+LDLIBS = -lpopart -lpoplar -lpopops
```
Comment (weimingzha0) [inline models/vision/detection/yolo/Makefile:10]: I think we don't need to link these libs. libodla_popart.so will link against these libs.

File: models/vision/detection/yolo/Makefile [hunk 8]
```diff
@@ -31,3 +32,9 @@ demo: ${out_dir}/demo ${out_dir}/test.jpg
 
 clean:
 	rm -fr ${out_dir}
+
+${out_dir}/yolo_postproc_ipu.o : yolov3_postproc_ipu.cc
+	g++ -std=c++14 -c ${flags} yolov3_postproc_ipu.cc ${LDLIBS} -ldl -DONNX_NAMESPACE=onnx -o $@ -D_GLIBCXX_USE_CXX11_ABI=1
```
Comment (weimingzha0) [inline models/vision/detection/yolo/Makefile:37]: use default ABI?

File: models/vision/detection/yolo/yolov3_postproc_ipu.cc [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+  decode<1, 13, 80>(orig_img_w, orig_img_h, bb13, anchor_masks[0]);
+  decode<1, 26, 80>(orig_img_w, orig_img_h, bb26, anchor_masks[1]);
+  decode<1, 52, 80>(orig_img_w, orig_img_h, bb52, anchor_masks[2]);
+  // NMS
+  float score_thre = 0.3;
+  float iou_thre = 0.45;
+  std::vector<std::pair<std::string, std::array<float, 5>>> ret;
+
+  void* handle =
+      dlopen("../ODLA/platforms/odla_popart/custom_ops/build/libcustom_ops.so",
```
Comment (weimingzha0) [inline models/vision/detection/yolo/yolov3_postproc_ipu.cc:122]: can we use libODLA_popart? In yolov3 code, we don't use IPU APIs directly. Instead, we ODLA apis.
We can construct a odla_computation and use odla_NMS() here.

**Pull request closed without merging**